consensus/src/consensus.rs

```diff
@@ -248,40 +276,76 @@ where
         .await;
 
         // Listen to incoming certificates.
-        while let Some(certificate) = self.rx_primary.recv().await {
-            let sequence =
-                self.protocol
-                    .process_certificate(&mut state, self.consensus_index, certificate)?;
-
-            // Update the consensus index.
-            self.consensus_index += sequence.len() as u64;
-
-            // Output the sequence in the right order.
-            for output in sequence {
-                let certificate = &output.certificate;
-                #[cfg(not(feature = "benchmark"))]
-                if output.consensus_index % 5_000 == 0 {
-                    tracing::debug!("Committed {}", certificate.header);
-                }
+        loop {
+            tokio::select! {
+                Some(certificate) = self.rx_primary.recv() => {
+                    // If the core already moved to the next epoch we should pull the next
+                    // committee as well.
+                    match certificate.epoch().cmp(&self.committee.epoch()) {
+                        Ordering::Greater => {
+                            let message = self.rx_reconfigure.borrow_and_update().clone();
+                            match message  {
+                                ReconfigureNotification::NewCommittee(new_committee) => {
+                                    state = self.reconfigure(new_committee)?;
+                                },
+                                ReconfigureNotification::Shutdown => return Ok(()),
+                            }
+                        }
+                        Ordering::Less => {
+                            // We already updated committee but the core is slow.
+                            tracing::debug!("Already moved to the next epoch");
+                            continue
+                        },
+                        Ordering::Equal => {
+                            // Nothing to do, we can proceed.
+                        }
```

Comment:
This is essentially the same as the proposer, right? There's probably a trait lurking under this.

Reply:
That could be nice indeed. I will think about it in a next PR

Reply:
That could be nice indeed. I will think about it in a next PR